### ISSUE TITLE
find venv of pyproject by default?

### ISSUE THREAD

**Issue by avivelmakias1** [linux-cultist/venv-selector.nvim#1]
Hi,
Most of my projects use `poetry`, and it seems that it doesn't detect it's venvs.
Is is possible to add a feature to detect the virtualenvs created by poetry?
I can't figure out how to make this package detect the venvs it's creating.

**Comment by linux-cultist** [comment]
Hi,

It's something I was thinking about adding actually. Right now the plugin just searches for venvs that have a fixed name (like venv or .venv).

Poetry seems to place all the venvs in one location, so the plugin could just let you pick one of those venvs.

I will add a parameter in the options where you can specify "poetry" and the plugin will look in the default poetry location.

It's very good that you opened this issue because I was thinking about adding it, but just wasn't sure if other people were using the plugin. It's quite new also.

Thank you for posting and I will see if I can add this in the coming days. :)

**Comment by avivelmakias1** [comment]
Really looking forward to see this!
Thanks you!

**Comment by linux-cultist** [comment]
Please update the plugin and check out the instructions for using it with Poetry in the README. :)

I hope it works for you, and if you have more ideas how to improve the plugin, im very open to hearing them!

**Comment by avivelmakias1** [comment]
looks like it's working perfect!
most of my are using `poetry` so its really nice.
if i had to think of something else that might add its going to be the option to to always search in the root dir of the project for common names like `venv`, so it'll work for both `poetry` and none `poetry` projects at the same time.

**Comment by avivelmakias1** [comment]
also i have another suggestion,
i'm also trying to configure debugging, and i see that this plugin doesn't add the current virtualenv that i picked into the `VIRTUAL_ENV` environment variable, which is happening when you activate a virtualenv.
setting this environment variable will be helpful to identify whether i'm in a venv or not for debugging.

**Comment by linux-cultist** [comment]
Will add that for sure. Thanks!

**Comment by linux-cultist** [comment]
I pushed a new version that will find Poetry and Pipenv environments by default without configuration.

Remove the extra parameters you needed to put in the setup function before when using Poetry. They are not needed anymore.

Try it out and see what you think!

I still havent added the environment variable `VIRTUAL_ENV` but its coming also. :)

**Comment by avivelmakias1** [comment]
looks like it got really slow for me, and not finding what i was looking for.
it feels like it's looking at my whole system.
i'm using mac, and the paths look a bit different from the ones you set as the defaults...
also it doesn't find my poetry venvs, just the file in the same path as the `virtualenvs`.
it's looking in other files that i didnt want to..

i didn't have much time to test it fully and figure out how to make it work again, but this is my config now and it's not working correctly... (using lazy.vim, i dont have pipenv installed)
```
{
    "linux-cultist/venv-selector.nvim",
    opts = {
      poetry_path = "/Users/<myusername>/Library/Caches/pypoetry/virtualenvs",
      -- name = "*",
      -- parents = 0,
      -- children = 0,
    },
    keys = { { "<leader>lv", "<cmd>:VenvSelect<cr>", desc = "Select VirtualEnv" } },
  }
```

**Comment by linux-cultist** [comment]
Let me set the default back to what it was, should fix it :) 

On Sat, Mar 18, 2023, at 20:23, Aviv Elmakias wrote:
> 
> 
> 
> looks like it got really slow for me, and not finding what i was looking for.
> it feels like it's looking at my whole system.
> i'm using mac, and the paths look a bit different from the ones you set as the defaults...
> also it doesn't find my poetry venvs, just the file in the same path as the `virtualenvs`.
> it's looking in other files that i didnt want to..
> 
> i didn't have much time to test it fully and figure out how to make it work again, but this is my config now and it's not working correctly... (using lazy.vim, i dont have pipenv installed)
> 
> `{
>     "linux-cultist/venv-selector.nvim",
>     opts = {
>       poetry_path = "/Users/<myusername>/Library/Caches/pypoetry/virtualenvs",
>       -- name = "*",
>       -- parents = 0,
>       -- children = 0,
>     },
>     keys = { { "<leader>lv", "<cmd>:VenvSelect<cr>", desc = "Select VirtualEnv" } },
>   }
`

**Comment by linux-cultist** [comment]
Try removing the name, parents and children parameters also, and just keep the poetry_path. This will set parents to 2, children is not used anymore, and name will be "venv" for your virtual environments that are not poetry (since they have dynamic names).

**Comment by linux-cultist** [comment]
![image](https://user-images.githubusercontent.com/126409588/226135993-69573931-706b-4b91-bee8-0c987defb883.png)

This is how it should look, with poetry and pipenv virtual environments included at the bottom. They are not included for you?

**Comment by avivelmakias1** [comment]
changing my config to this:
```
{
    "linux-cultist/venv-selector.nvim",
    opts = {
      poetry_path = "/Users/<myusername>/Library/Caches/pypoetry/virtualenvs",
      -- name = "*",
      parents = 0,
    },
    keys = { { "<leader>lv", "<cmd>:VenvSelect<cr>", desc = "Select VirtualEnv" } },
  }
```
made it faster, but it's not recognizing my venvs now, just the `envs.toml`  file inside here: "/Users/<myusername>/Library/Caches/pypoetry/virtualenvs"

**Comment by avivelmakias1** [comment]
maybe adding `-t directory` to the `fd` command will eliminate showing me `envs.toml` file inside my `pypoetry/virtualenvs` directory.
but still it doesn't tell me why i can't see my actual venvs

**Comment by linux-cultist** [comment]
I think you are on to something for sure, and I only used one virtual environment to test with, so I didn't notice the bug.

Will fix tomorrow :) Thank you very much!

**Comment by linux-cultist** [comment]
I could replicate exactly what you were seeing and fix it, please try latest version now. :)

**Comment by avivelmakias1** [comment]
looks like it's working again.
i had to keep the `parents = 0` parameter to be fast, without it it's working but showing me other directories which aren't venvs.
maybe by default it should be set to 0 so it'll look just inside the root dir and the default dirs for poetry and pipenv.

now i just need the env variables and everything will be great :)

**Comment by linux-cultist** [comment]
Yeah the search for venvs outside of Poetry or Pipenv takes some time, and if you only use Poetry or Pipenv, its also completely unnecessary for it to search. :)

I will put in an option you can set to not search outside of Poetry or Pipenv locations, which will make it very fast.

Will look at adding the environment variables next. :)

Thanks for using the plugin and reporting bugs, I really appreciate it!

**Comment by avivelmakias1** [comment]
its showing me more paths with the word `venv` in them, but not things that i'm using (like venvs of mason)
setting `parents` to 0 fix it for me so i'm fine with it like this.

**Comment by linux-cultist** [comment]
The plugin will now add the selected virtual environment to `PATH` and also set `VIRTUAL_ENV` environment variable. So if you open a new terminal from neovim after selecting a virtual environment, you should see it being activated in the terminal as well.

Let me know how it works for you. :)

**Comment by avivelmakias1** [comment]
hi,
i checked it and it's not quite right like using `activate`.
the environment variable `VIRTUAL_ENV` should just be the path to the venv, not the path to `bin/python` inside it.

also, opening a new terminal will set the `bin` folder after other stuff that i exported (like pyenv or other stuff) but that's something that i need to figure out on my own (it's happening because i exported them in my `.zshrc` file)

**Comment by avivelmakias1** [comment]
looks like something else broke.
now the pyright lsp isn't working.
this is the message i get when i select a venv:
" Pyright now using '/Users/<myusername>/Library/Caches/pypoetry/virtualenvs/<myproject>/Users/<myusername>/Library/Caches/pypoetry/virtualenvs/<myproject>/bin/python'."
looks like it's the path to the venv + the path to `bin/python` in the venv combined

**Comment by linux-cultist** [comment]
I really need to add tests to this code base to catch stupid mistakes like this.

Fixed. :)

**Comment by linux-cultist** [comment]
Very good, im really happy to hear that! 

Huge thanks for all the help, and I will add tests to the code to make sure I dont break anything on future features. :)
